zebra-state/src/service/read/difficulty.rs

```
@@ -0,0 +1,62 @@
+//! Get context and calculate difficulty for the next block.
+//!
+use std::borrow::Borrow;
+
+use zebra_chain::{
+    block::{Block, Hash, Height},
+    parameters::{Network, POW_AVERAGING_WINDOW},
+    work::difficulty::{CompactDifficulty, ExpandedDifficulty, U256},
+};
+
+use crate::service::{
+    any_ancestor_blocks,
+    check::{difficulty::POW_MEDIAN_BLOCK_SPAN, AdjustedDifficulty},
+    finalized_state::ZebraDb,
+    NonFinalizedState,
+};
+
+/// Return the CompactDifficulty for the current best chain.
+///
+/// Note: Return `1` as the difficulty if we don't have enough blocks in the state. Should not happen in a
+/// running blockchain but only in some test cases where not enough state is loaded.
+pub fn relevant_chain_difficulty(
+    non_finalized_state: NonFinalizedState,
+    db: &ZebraDb,
+    tip: (Height, Hash),
+    network: Network,
+) -> CompactDifficulty {
+    let relevant_chain = any_ancestor_blocks(&non_finalized_state, db, tip.1);
+    difficulty(relevant_chain, tip.0, network)
+}
+
+fn difficulty<C>(relevant_chain: C, tip_height: Height, network: Network) -> CompactDifficulty
+where
+    C: IntoIterator,
+    C::Item: Borrow<Block>,
+    C::IntoIter: ExactSizeIterator,
+{
+    const MAX_CONTEXT_BLOCKS: usize = POW_AVERAGING_WINDOW + POW_MEDIAN_BLOCK_SPAN;
+
+    let relevant_chain: Vec<_> = relevant_chain
+        .into_iter()
+        .take(MAX_CONTEXT_BLOCKS)
+        .collect();
+
+    let relevant_data = relevant_chain.iter().map(|block| {
+        (
+            block.borrow().header.difficulty_threshold,
+            block.borrow().header.time,
+        )
+    });
+
+    // If we don't have enough context we just return 1.
+    if relevant_data.len() < MAX_CONTEXT_BLOCKS {
+        return CompactDifficulty::from(ExpandedDifficulty::from(U256::one()));
```

Zebra will panic here if there is not enough context. I made this to make tests pass but i am unsure if it will be the best way to handle.

---

I think we should return an error saying there's not enough context and use a MockService to return this in the test